### PR TITLE
[Markdown] Add support for LaTeX math blocks

### LINKED ISSUE #3251: [Markdown] Highlight of $$...$$ math blocks
### What happened?

1.    Expected two separated highlighted blocks:
```
This is $x_1$ and $y_2$.
```
2.    Expected two separated highlighted blocks:
```
This is $$x_1$$ and $$y_2$$.
```
3.    Expected highlighted as math block:
```
$$
c=\sqrt{a^2+b^2}
$$
```

@cser FYI.

### PR DESCRIPTION
**Pull request by deathaxe**
Closes #3251

### CODE REVIEW

**Review by jrappen — DISMISSED** [review]

**Review by jrappen — DISMISSED** [review]

File: Markdown/tests/syntax_test_markdown.md [hunk 1]
```diff
@@ -7867,3 +7867,63 @@ This is a [[wiki link]].
 |         ^^^^^^^^^^^^^ meta.link.reference.wiki.description.markdown
 |         ^^ punctuation.definition.link.begin.markdown
 |                    ^^ punctuation.definition.link.end.markdown
+
+
+# TEST: LATEX EQUATIONS #######################################################
```
Comment (FichteFoll) [inline Markdown/tests/syntax_test_markdown.md:7872]: Please add a test for an unclosed inline math expression that is not highlighted.

And one for the requirement of no whitespace on the inner boundaries of the enclosing punctuation, i.e. `$ 1 + 3 $`.

**Review by FichteFoll — CHANGES_REQUESTED** [review]

File: Markdown/Markdown.sublime-syntax [hunk 2]
```diff
@@ -2895,6 +2906,50 @@ contexts:
...
+###[ EXTENSIONS: MATHJAX ]####################################################
+
+  math-blocks:
+    - match: '[ \t]*(\$\$)'
+      captures:
+        1: punctuation.definition.math.begin.markdown
+      push: math-block-content
+
+  math-block-content:
+    - meta_scope:
```
[the remark this answers is not in the document]
Comment (deathaxe) [reply Markdown/Markdown.sublime-syntax:2918]: Maybe we need to distinguish those blocks with fenced-codeblock style standalone `$$` and those which look like inlines `$$ 1+1 $$`.

[the remark this answers is not in the document]
Comment (jwortmann) [reply Markdown/Markdown.sublime-syntax:2918]: +1 for treating fenced-codeblock style `$$` and inline-style `$$` LaTeX blocks differently.

Interestingly, rendering on GitHub doesn't follow it's own specs as described in the [GitHub docs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions#writing-expressions-as-blocks):

> To add a math expression as a block, **start a new line** and delimit the expression with two dollar symbols `$$`.

but it is also rendered as a LaTeX *block* on GitHub if there is preceding text (i.e. no new line) and following text:

```md
Some text $$\sqrt{2\pi}$$ more text
```
Some text $$\sqrt{2\pi}$$ more text

... while this PR recognizes it as inline math

pic1

**Comment by jwortmann** [comment]
I wonder whether the following can be improved ([according to docs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions#writing-dollar-signs-in-line-with-and-within-mathematical-expressions), `span` tags should disable parsing the `$` as LaTeX delimiter):

```md
<span>$</span>100 this is not LaTeX math \sqrt{2\pi} <span>$</span>200
```

Rendered on GitHub as:

<span>$</span>100 this is not LaTeX math \sqrt{2\pi} <span>$</span>200

Highlighting from this PR:

pic2

Perhaps the `<span>` tag should be treated specially, to ignore the LaTeX delimiter? Or maybe would it be enough to adjust the priority of included contexts? (I haven't really looked at the implementation)

Interestingly, the highlighting is not affected if the 2nd `$` isn't enclosed in a `span` tag:

pic3

**Comment by deathaxe** [comment]
Current PR is part of MDE for years without any complains. I am uncertain about how much effort we should invest into such uncommon odd edge cases with regards to resulting complexity readability. Starting to push contexts for inline html tags is very likely to not play well with emphasises at some point. Things get worth if we want to fail on paragraph ends, I guess.

**Review by jrappen — DISMISSED** [review]

**Comment by deathaxe** [comment]
Maybe worth digging into it a bit deeper to improve some edge cases.

**Comment by jwortmann** [comment]
Pretty neat that the last commit solves the problem with `<span>$</span>`, but I fear to reject everything that looks like a HTML tag is not the best approach. The following example with a "smaller than" operator breaks syntax highlighting further down the file (I would say it is not very uncommon to have no whitespace around operators or other symbols in LaTeX):

```markdown
Some text with $a<b$ inline math

Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non
proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
```

Unbenannt

Rendering on GitHub does still have some similar issues as well: https://github.com/github/feedback/discussions/17022

**Comment by deathaxe** [comment]
I guess we could play that game forever. There are just too many ambiguities in Markdown, which can barely handled by a lexer which needs to get all things done in one cycle.

Just consider start or end of bold/italics in the middle of equations, which causes them to not be rendered. `*text $1*2$` (*text $1*2$). I doubt we want/can handle them all.

**Comment by jwortmann** [comment]
True. But I guess the more of critical edge cases which may break syntax highlighting we can find and improve now, the less bug reports will appear after this is merged and released.

**Comment by michaelblyons** [comment]
> Just consider start or end of bold/italics in the middle of equations, which causes them to not be rendered. `*text $1*2$` (_text $1_2$). I doubt we want/can handle them all.

And furthermore, I doubt they are interpreted consistently in parsers that _do_ support MathJax.

**Comment by deathaxe** [comment]
MathJax expressions are rendered by browser. Each markdown parser just handles it as plain text thus bold/italics/htmltags have higher priority - each causing the equation to render invalid.

Main issue is it being hard up to impossible to immitate those multiple cycles of parsing markdown in a single run using ST's syntax engine.

File: Markdown/Markdown.sublime-syntax (hunk 2 above)
[the remark this answers is not in the document]
Comment (FichteFoll) [reply Markdown/Markdown.sublime-syntax:2919]: Thanks for the clarification. Imo, if inline `$$` is supported in some fashion (like it apparently is by GitHub), we should have separate matchings for the inline and the block version, but if it's not exactly documented then it's not a requirement from my side. Feel free to resolve if deemed unnecessary.

**Comment by deathaxe** [comment]
Synced source with MarkdownEditing 3.1.5

File: Markdown/Markdown.sublime-syntax (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwortmann) [reply Markdown/Markdown.sublime-syntax:3014]: Looks like this has been implemented and can be marked as resolved now.

**Comment by jwortmann** [comment]
Is it ready for review?

**Comment by FichteFoll** [comment]
Unfortunately, the force-pushes make the PR rather hard to follow as it isn't as trivial to immediately spot the new changes in the new commits. GitHub provides a comparison link for the previous and new commits, but especially when merge commits or rebases are involved, it becomes non-trivial to detect these.

Since the PRs are getting squashed anyway, I believe it would be best to not rebase and force-push PRs instead of merging in master and adding more commits with fixes, especially when in response to review comments.

**Review by FichteFoll — DISMISSED** [review]

**Comment by deathaxe** [comment]
I started with mergin master in, but found changes significant enough (due to branching) to be easier looked at as a whole. Sure context to previous changes was lost.

File: Markdown/Markdown.sublime-syntax [hunk 3]
```diff
@@ -2956,6 +2960,172 @@ contexts:
...
+
+  math-inline:
+    # Consume any remaining $$ to prevent deadlock
+    # (workaround for sublimehq/sublime_text/issues/5415)
+    - match: \$\$
+    - match: |-
+        (?x)
+        # dollar sign not followed by whitespace
+        (\$) (?=\S)
+        # arbritary number of chars, maybe escaped dollar sign, but no html tags
```
Comment (jrappen) [inline Markdown/Markdown.sublime-syntax:3105]: typo

File: Markdown/Markdown.sublime-syntax [hunk 4]
```diff
@@ -2956,6 +2960,172 @@ contexts:
...
+  math-block-content:
+    - meta_include_prototype: false
+    - match: ''
+      embed: math-content
+      escape: (?=\$\$)|{{paragraph_end}}|{{setext_heading_escape}}
+      pop: 1 # embed math once
+
+  math-inline:
+    # Consume any remaining $$ to prevent deadlock
+    # (workaround for sublimehq/sublime_text/issues/5415)
```
Comment (jrappen) [inline Markdown/Markdown.sublime-syntax:3099]: if you put the full URL, we can just right-click & open URL in browser from within Sublime

**Comment by jrappen** [comment]
Code seems fine from a once-over. Comments had typos.

**Review by jrappen — APPROVED** [review]

**Review by FichteFoll — APPROVED** [review]

**Review by michaelblyons — APPROVED** [review]
